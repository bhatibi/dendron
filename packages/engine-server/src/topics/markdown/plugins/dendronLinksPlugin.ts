import RemarkParser, { Eat } from "remark-parse";
import { Processor, Parser } from "unified";
import { Node } from "unist";
import { removeMDExtension } from "@dendronhq/common-server";
import { DNoteLoc } from "@dendronhq/common-all";

const LINK_REGEX = /^\[\[(.+?)\]\]/;

declare class RemarkParserClass implements Parser {
  Parser: RemarkParserClass;
  prototype: RemarkParserClass;
  parse(): Node;
  blockMethods: string[];
  inlineTokenizers: {
    [key: string]: RemarkParser.Tokenizer;
  };

  inlineMethods: string[];
}

export type WikiLinkNote = Node & {
  data: WikiLinkData;
  value: string;
};

export type WikiLinkData = {
  alias: string;
  permalink: string;
  exists: boolean;
  hName: string;
  hProperties: any;
  hChildren: any[];

  // dendron specific
  toMd?: boolean;
  toHTML?: boolean;
  prefix?: string;
  useId: boolean;
  note?: { id: string };
  replace?: DNoteLoc;
  forNoteRefInPreview?: boolean;
  forNoteRefInSite?: boolean;
};

function locator(value: string, fromIndex: number) {
  return value.indexOf("[", fromIndex);
}

interface PluginOpts {
  permalinks?: string[];
  pageResolver?: (name: string) => string[];
  newClassName?: string;
  wikiLinkClassName?: string;
  hrefTemplate?: (permalink: string) => string;
  aliasDivider?: string;
  replaceLink?: { from: DNoteLoc; to: DNoteLoc };
}

export { PluginOpts as DendronLinksOpts };

/**
 * Matches wiki-links
 * @param opts
 */
export function dendronLinksPlugin(opts: Partial<PluginOpts> = {}) {
  const permalinks = opts.permalinks || [];
  const defaultPageResolver = (name: string) => [
    name.replace(/ /g, "").toLowerCase(),
  ];
  const pageResolver = opts.pageResolver || defaultPageResolver;
  const newClassName = opts.newClassName || "new";
  const wikiLinkClassName = opts.wikiLinkClassName || "internal";
  const defaultHrefTemplate = (permalink: string) => `#/page/${permalink}`;
  const hrefTemplate = opts.hrefTemplate || defaultHrefTemplate;
  const aliasDivider = opts.aliasDivider || "|";

  function isAlias(pageTitle: string) {
    return pageTitle.indexOf(aliasDivider) !== -1;
  }

  function parseAliasLink(pageTitle: string) {
    const [displayName, name] = pageTitle.split(aliasDivider);
    return { name, displayName };
  }

  function parsePageTitle(pageTitle: string) {
    if (isAlias(pageTitle)) {
      return parseAliasLink(pageTitle);
    }
    return {
      name: pageTitle,
      displayName: pageTitle,
    };
  }

  function inlineTokenizer(eat: Eat, value: string) {
    const match = LINK_REGEX.exec(value);

    if (match) {
      const pageName = match[1].trim();
      const { name, displayName } = parsePageTitle(pageName);

      const pagePermalinks = pageResolver(name);
      let permalink = pagePermalinks.find((p) => permalinks.indexOf(p) !== -1);
      const exists = permalink !== undefined;

      if (!exists) {
        permalink = pagePermalinks[0];
      }

      let classNames = wikiLinkClassName;
      if (!exists) {
        classNames += " " + newClassName;
      }
      // normalize
      if (permalink) {
        permalink = removeMDExtension(permalink);
      }

      return eat(match[0])({
        type: "wikiLink",
        value: name,
        data: {
          alias: displayName,
          permalink,
          exists,
          hName: "a",
          hProperties: {
            className: classNames,
            href: hrefTemplate(permalink as string),
          },
          hChildren: [
            {
              type: "text",
              value: displayName,
            },
          ],
        },
      });
    }
    return;
  }

  inlineTokenizer.locator = locator;

  // @ts-ignore
  let _this = (this as any) as RemarkParserClass;
  const Parser: RemarkParser.Parser = _this.Parser;
  // @ts-ignore
  const inlineTokenizers = Parser.prototype.inlineTokenizers;
  // @ts-ignore
  const inlineMethods = Parser.prototype.inlineMethods;
  inlineTokenizers.wikiLink = inlineTokenizer;
  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "wikiLink");

  // Stringify for wiki link
  // @ts-ignore
  const Compiler: Processor["Compiler"] = _this.Compiler;

  if (Compiler != null) {
    const visitors = Compiler.prototype.visitors;
    if (visitors) {
      visitors.wikiLink = function (node: Node) {
        const data = node.data as WikiLinkData;
        if (!node || !node.data || !node.data.alias) {
          throw Error(`no alias found: ${JSON.stringify(node)}`);
        }
        if (data.useId) {
          if (!data.note) {
            throw Error("no note");
          }
          node.value = data.note.id;
        }
        if (opts.replaceLink && opts.replaceLink.from.fname === node.value) {
          // TODO: check for case
          node.value = opts.replaceLink.to.fname;
          if (node.data.alias === opts.replaceLink.from.fname) {
            node.data.alias = opts.replaceLink.to.fname;
          }
        }
        if (data.toMd) {
          return `[${data.alias}](${data.prefix || ""}${node.value})`;
        }
        if (data.forNoteRefInPreview) {
          return `${data.prefix || ""}${node.value}.md`;
        }
        if (data.forNoteRefInSite) {
          return `${data.prefix || ""}${node.value}.html`;
        }
        if (data.toHTML) {
          return `${data.prefix || ""}${node.value}.html`;
        }

        if (node.data.alias !== node.value) {
          return `[[${node.data.alias}${aliasDivider}${node.value}]]`;
        }
        return `[[${node.value}]]`;
      };
    }
  }
}
