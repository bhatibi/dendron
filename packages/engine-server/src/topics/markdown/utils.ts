/* eslint-disable camelcase */
import { DNodeUtils, LinkType, Note, ProtoLink } from "@dendronhq/common-all";
import fs from "fs-extra";
import _ from "lodash";
import _markdownit from "markdown-it";
import markdownItRegex from "markdown-it-regex";
import Token from "markdown-it/lib/token";
import os from "os";
import remark from "remark";
import abbrPlugin from "remark-abbr";
import frontmatterPlugin from "remark-frontmatter";
import markdownParse from "remark-parse";
// import wikiLinkPlugin from "remark-wiki-link";
import unified, { Processor } from "unified";
import { Node, Parent, Point, Position } from "unist";
import visit, { CONTINUE, EXIT } from "unist-util-visit";
import YAML from "yamljs";
import { dendronLinksPlugin } from "./plugins/dendronLinksPlugin";
import { dendronRefsPlugin } from "./plugins/dendronRefsPlugin";
import { ReplaceRefOptions } from "./plugins/replaceRefs";
import { MDRenderer } from "./renderer";

enum NodeType {
  code_inline = "code_inline",
  code_block = "code_block",
  fence = "fence",
  image = "image",
  hardbreak = "hardbreak",
  softbreak = "softbreak",
  text = "text",
  html_block = "html_block",
  html_inline = "html_inline",
}

let processor: unified.Processor | null = null;

export type MDNode = {
  nodeType: NodeType;
  openNode: Token;
  closeNode: Token;
  children: Token[];
};

export function md() {
  return _markdownit();
}

export function tokens2MD(tokens: Token[]): string {
  return _.trim(new MDRenderer().render(tokens, {}, {}));
}

export function md2Tokens(txt: string): Token[] {
  const markdownIt = _markdownit();
  const tokens: Token[] = markdownIt
    .use(markdownItRegex, {
      name: "ref-document",
      // eslint-disable-next-line no-useless-escape
      regex: /\[\[([^\[\]]+?)\]\]/,
      // replace: (rawRef: string) => {
      //   const { ref, label } = parseRef(rawRef);
      //   // const fsPath = findUriByRef(getWorkspaceCache().allUris, ref)?.fsPath;
      //   // if (!fsPath) {
      //   //   return getInvalidRefAnchor(label || ref);
      //   // }
      //   return getRefAnchor(ref, label || ref);
      // },
    })
    .parse(txt, {});
  return tokens;
}

// export function replaceMDRefs(opts: ReplaceMDRefsOpts): Token[] {
//   const { tokens, imageLinkPrefix } = opts;
//   const reImg = new RegExp("((?<link>.*))");
//   tokens.map((t) => {
//     if (t.type === "inline") {
//       const children: Token[] = t.children ? t.children : [];
//       if (children[0]?.type === "image" && imageLinkPrefix) {
//         const attrs = children[0].attrs;
//         const oldSrcAtt = _.find(attrs, (u) => u[0] === "src");
//         const [_skip, oldSrc] = oldSrcAtt;
//         const newSrc = imageLinkPrefix + oldSrc;
//         // const match = reImg.exec(t.content);
//         // if (!match || !match.groups) {
//         //   throw Error(`no image found for token ${t}`);
//         // }
//         // const newLink = imageLinkPrefix + match.groups["link"];
//         t.content = t.content.replace(`(${oldSrc})`, `(${newSrc})`);
//       }
//     }

//     //   (!_.isEmpty(t.children)) &&
//     //   (t.children[0].type === "image") &&
//     //   imageLinkPrefix
//     // ) {
//     // }
//   });
//   return tokens;
// }

export function parse(markdown: string): Node {
  processor =
    processor ||
    unified()
      .use(markdownParse, { gfm: true })
      .use(abbrPlugin)
      .use(frontmatterPlugin, ["yaml"])
      .use(dendronLinksPlugin);
  return processor.parse(markdown);
}

// export async function process(node: Node) {
//   const streamIn = new PassThrough();
//   streamIn.write("# foo bar");

//   return new Promise((resolve) => {
//     engine(
//       {
//         processor: remark(),
//         streamIn,
//         pluginPrefix: "remark",
//         rcName: ".remarkrc",
//         packageField: "remarkConfig",
//         ignoreName: ".remarkignore",
//         color: true,
//       },
//       // eslint-disable-next-line no-use-before-define
//       done
//     );

//     function done(error: any) {
//       if (error) throw error;
//       resolve();
//     }
//   });
// }

// TODO: stub
function uriToSlug(uri: string) {
  return uri;
}

export function createNoteFromMarkdown(uri: string, eol?: string): Note {
  // eslint-disable-next-line prefer-const
  const markdown = fs.readFileSync(uri, { encoding: "utf8" });
  eol = eol || os.EOL;

  const tree = parse(markdown);
  let title: string | null = null;
  let frontmatter: any = {};

  visit(tree, (node) => {
    if (node.type === "heading" && node.depth === 1) {
      if (_.isNull(title)) {
        title = ((node as Parent)!.children[0].value as string) || title;
      }
    }
    return title === null ? CONTINUE : EXIT;
  });

  const links: ProtoLink[] = [];
  let start: Point = { line: 1, column: 1, offset: 0 }; // start position of the note

  visit(tree, (node) => {
    if (node.type === "yaml") {
      frontmatter = YAML.parse(node.value as string) ?? {}; // parseYAML returns null if the frontmatter is empty
      if (frontmatter.tite) {
        title = frontmatter.title;
      }
      // Update the start position of the note by exluding the metadata
      start = {
        line: node.position!.end.line! + 1,
        column: 1,
        offset: node.position!.end.offset! + 1,
      };
    }

    if (node.type === "image") {
      links.push({
        type: LinkType.IMAGE_LINK,
        url: node.url as string,
        alt: node.alt as string,
        position: node.position!,
      });
    }

    if (node.type === "wikiLink") {
      let maybeLbl: string | undefined;
      let maybeUrl: string | undefined;
      [maybeLbl, maybeUrl] = (node.value as string).split("|");
      if (_.isUndefined(maybeUrl)) {
        maybeUrl = maybeLbl;
        maybeLbl = undefined;
      }
      links.push({
        type: LinkType.WIKI_LINK,
        position: node.position!,
        url: maybeUrl,
      });
    }

    // if (node.type === "definition") {
    //   linkDefinitions.push({
    //     label: node.label as string,
    //     url: node.url as string,
    //     title: node.title as string,
    //     position: node.position,
    //   });
    // }
  });

  // Give precendence to the title from the frontmatter if it exists
  // title = frontmatter.title ?? title;

  const end = tree.position!.end;
  // const definitions = getFoamDefinitions(linkDefinitions, end);

  const props = {
    properties: frontmatter,
    slug: uriToSlug(uri),
    title,
    links,
    source: {
      tree,
      uri,
      text: markdown,
      contentStart: start,
      end,
      eol,
    },
  };
  const fname = DNodeUtils.basename(props.source.uri, true);
  const note = new Note({ fname });
  note.custom.props = props;
  return note;
}

// === Text Edits

// export const generateLinkPrefix = (
//   note: Note,
//   links: ProtoLink[],
//   prefix: string
// ): string => {
//   const imageRe = new RegExp(/(\([^)]+\))/);
//   let text = [];
//   links.map((l) => {
//     const source = NoteUtils.protoGetSource(note);
//     const { position, url } = l;
//     // const match = imageRe.exec(url);
//     // if (!match) {
//     //   throw Error("bad image url");
//     // }
//     const { start, end } = position;
//     const oldLink = source.text.slice(start.offset, end.offset);
//     const newUrl = [prefix, url].join("");
//     // const capture = match[1];
//     const newLink = oldLink.replace(`(${url})`, `(${newUrl})`);
//     console.log("foo");
//     // eslint-disable-next-line no-use-before-define
//     const newText = applyTextEdit(source.text, {
//       newText: newLink,
//       range: {
//         start,
//         end,
//       },
//     });
//   });
//   return [];
// };

// export const generateHeading = (note: Note): TextEdit | null => {
//   if (!note) {
//     return null;
//   }

//   if (note.title) {
//     return null;
//   }

//   const frontmatterExists = note.source.contentStart.line !== 1;

//   let newLineExistsAfterFrontmatter = false;
//   if (frontmatterExists) {
//     const lines = note.source.text.split(note.source.eol);
//     const index = note.source.contentStart.line - 1;
//     const line = lines[index];
//     newLineExistsAfterFrontmatter = line === "";
//   }

//   const paddingStart = frontmatterExists ? note.source.eol : "";
//   const paddingEnd = newLineExistsAfterFrontmatter
//     ? note.source.eol
//     : `${note.source.eol}${note.source.eol}`;

//   return {
//     newText: `${paddingStart}# ${getHeadingFromFileName(
//       note.slug
//     )}${paddingEnd}`,
//     range: {
//       start: note.source.contentStart,
//       end: note.source.contentStart,
//     },
//   };
// };

export interface TextEdit {
  range: Position;
  newText: string;
}

export const applyTextEdit = (text: string, textEdit: TextEdit): string => {
  const characters = text.split("");
  const startOffset = textEdit.range.start.offset || 0;
  const endOffset = textEdit.range.end.offset || 0;
  const deleteCount = endOffset - startOffset;

  const textToAppend = `${textEdit.newText}`;
  characters.splice(startOffset, deleteCount, textToAppend);
  return characters.join("");
};

export function getProcessor(opts?: {
  root?: string;
  renderWithOutline?: boolean;
  replaceRefs?: ReplaceRefOptions;
}): Processor {
  const { root, renderWithOutline, replaceRefs } = opts || {};
  return remark()
    .use(markdownParse, { gfm: true })
    .use(abbrPlugin)
    .use(frontmatterPlugin, ["yaml"])
    .use(dendronLinksPlugin)
    .use(dendronRefsPlugin, { root, renderWithOutline, replaceRefs })
    .use({ settings: { listItemIndent: "1", fences: true } });
}
