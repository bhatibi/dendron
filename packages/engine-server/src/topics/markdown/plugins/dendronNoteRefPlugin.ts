import { DendronError, DEngineClientV2, Note } from "@dendronhq/common-all";
import { removeMDExtension } from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import { Parent } from "mdast";
import path from "path";
import { Eat } from "remark-parse";
import { Processor } from "unified";
import { Node } from "unist";
import { DendronRefLink, parseDendronRef } from "../../../utils";
import { ParserUtilsV2 } from "../utilsv2";
import { findIndex, isHeading } from "./inject";
import { ReplaceRefOptions, replaceRefs } from "./replaceRefs";

const LINK_REGEX = /^\(\((?<ref>[^)]+)\)\)/;

type CompilerOpts = {
  engine: DEngineClientV2;
  renderWithOutline: boolean;
  replaceRefOpts: ReplaceRefOptions;
  refLvl?: number;
};

type RefLinkData = {
  link: DendronRefLink;
  // --- Old
  alias: string;
  permalink: string;
  exists: boolean;
  hName: string;
  hProperties: any;
  hChildren: any[];
  toMd?: boolean;
  prefix?: string;
  useId: boolean;
  note?: Note;
};
// --- Utils
function doRenderWithOutline(opts: {
  content: string;
  title: string;
  link: string;
}) {
  const { content, title, link } = opts;
  return `

<div class="portal-container">
<div class="portal-head">
<div class="portal-backlink" >
<div class="portal-title">From <span class="portal-text-title">${title}</span></div>
<a href="${link}" class="portal-arrow">Go to text <span class="right-arrow">â†’</span></a>
</div>
</div>
<div id="portal-parent-anchor" class="portal-parent" markdown="1">
<div class="portal-parent-fader-top"></div>
<div class="portal-parent-fader-bottom"></div>        
  
${content}

</div>    
</div>
`;
}

const MAX_REF_LVL = 2;

function extractNoteRef(opts: {
  body: string;
  link: DendronRefLink;
  engine: DEngineClientV2;
  renderWithOutline: boolean;
  replaceRefOpts: ReplaceRefOptions;
  refLvl: number;
}) {
  const {
    body,
    link,
    engine,
    renderWithOutline,
    replaceRefOpts,
    refLvl,
  } = opts;

  const proc = ParserUtilsV2.getRemark().use(plugin, {
    engine,
    renderWithOutline,
    replaceRefOpts,
    refLvl,
  });

  // parse it so we can slice it
  const bodyAST: Parent = proc.parse(body) as Parent;
  const { anchorStart, anchorEnd, anchorStartOffset } = link;
  let anchorStartIndex = bodyAST.children[0].type === "yaml" ? 1 : 0;
  let anchorEndIndex = bodyAST.children.length;
  if (anchorStart) {
    anchorStartIndex = findIndex(bodyAST.children as any[], function (
      node: Node
    ) {
      return isHeading(node, anchorStart);
    });
    if (anchorStartIndex < 0) {
      return proc.stringify(
        ParserUtilsV2.genMDError({
          msg: `start anchor ${anchorStart} not found`,
          title: "Note Ref Error",
        })
      );
    }
  }
  if (anchorEnd) {
    anchorEndIndex = findIndex(
      bodyAST.children.slice(anchorStartIndex + 1) as any[],
      function (node: Node) {
        return isHeading(node, anchorEnd);
      }
    );
    if (anchorEndIndex < 0) {
      const mdError = ParserUtilsV2.genMDError({
        msg: `end anchor ${anchorEnd} not found`,
        title: "Note Ref Error",
      });
      return proc.stringify(mdError);
    }
    anchorEndIndex += anchorStartIndex + 1;
  }
  bodyAST.children = bodyAST.children.slice(anchorStartIndex, anchorEndIndex);

  // take the output and convert it using the full md toolchain
  let outProc = ParserUtilsV2.getRemark()
    .use(plugin, opts)
    .use(replaceRefs, replaceRefOpts);
  try {
    let out = outProc.processSync(outProc.stringify(bodyAST)).toString();
    if (anchorStartOffset) {
      out = out.split("\n").slice(anchorStartOffset).join("\n");
    }
    return out;
  } catch (err) {
    console.log(err);
    throw err;
  }
}

// --- Main

function plugin(opts: CompilerOpts) {
  // @ts-ignore
  let _this: Processor = this;
  attachParser({ proc: _this });
  attachCompiler({ proc: _this, ...opts });
}

function attachParser(opts: { proc: Processor }) {
  const { proc } = opts;
  const permalinks: any[] = [];
  const aliasDivider = "|";

  function isAlias(pageTitle: string) {
    return pageTitle.indexOf(aliasDivider) !== -1;
  }
  const pageResolver = (name: string) => [name.replace(/ /g, "").toLowerCase()];

  function locator(value: string, fromIndex: number) {
    return value.indexOf("((", fromIndex);
  }

  function parsePageTitle(pageTitle: string) {
    if (isAlias(pageTitle)) {
      return parseAliasLink(pageTitle);
    }
    return {
      name: pageTitle,
      displayName: pageTitle,
    };
  }
  function parseAliasLink(pageTitle: string) {
    const [displayName, name] = pageTitle.split(aliasDivider);
    return { name, displayName };
  }

  function inlineTokenizer(eat: Eat, value: string) {
    const match = LINK_REGEX.exec(value);

    if (match) {
      const pageName = match[1].trim();
      const { link } = parseDendronRef(pageName);
      const { name, displayName } = parsePageTitle(pageName);

      const pagePermalinks = pageResolver(name);
      let permalink = pagePermalinks.find((p) => permalinks.indexOf(p) !== -1);
      const exists = permalink !== undefined;

      if (!exists) {
        permalink = pagePermalinks[0];
      }

      let classNames = "internal";
      if (!exists) {
        classNames += " " + "new";
      }
      // normalize
      if (permalink) {
        permalink = removeMDExtension(permalink);
      }

      return eat(match[0])({
        type: "refLink",
        value: name,
        data: {
          alias: displayName,
          link,
          permalink,
          exists,
          hName: "a",
          hProperties: {
            className: classNames,
            href: permalink,
          },
          hChildren: [
            {
              type: "text",
              value: displayName,
            },
          ],
        },
      });
    }
    return;
  }

  inlineTokenizer.locator = locator;

  const Parser = proc.Parser;
  const inlineTokenizers = Parser.prototype.inlineTokenizers;
  const inlineMethods = Parser.prototype.inlineMethods;
  inlineTokenizers.refLink = inlineTokenizer;
  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "refLink");
  return Parser;
}

function attachCompiler(
  opts: {
    proc: Processor;
  } & CompilerOpts
) {
  const {
    refLvl,
    proc,
    engine,
    renderWithOutline,
    replaceRefOpts,
  } = _.defaults(opts, { refLvl: 0 });

  const Compiler = proc.Compiler;
  const visitors = Compiler.prototype.visitors;
  visitors.refLink = function (node: Node) {
    if (refLvl >= MAX_REF_LVL) {
      return "ERROR: Too many nested note references";
    }
    const data = node.data as RefLinkData;
    const root = engine.vaults[0];
    let body: string;
    try {
      body = fs.readFileSync(path.join(root, data.link.name + ".md"), {
        encoding: "utf8",
      });
    } catch (err) {
      const errors = proc.data("errors") as DendronError[];
      const msg = `${data.link.name} not found`;
      errors.push(new DendronError({ msg }));
      return proc.stringify(
        ParserUtilsV2.genMDError({
          msg,
          title: "Note Ref Error",
        })
      );
    }
    const out = extractNoteRef({
      body,
      link: data.link,
      engine,
      renderWithOutline,
      replaceRefOpts,
      refLvl: refLvl + 1,
    });
    if (renderWithOutline) {
      let link = data.link.name;
      link = _.trim(
        ParserUtilsV2.getRemark()
          .use(plugin, {
            engine,
            renderWithOutline,
            replaceRefOpts,
            refLvl,
          })
          .use(replaceRefs, { ..._.omit(replaceRefOpts, "wikiLink2Md") })
          .processSync(`[[${link}]]`)
          .toString()
      );
      return doRenderWithOutline({
        content: out,
        title: data.link.label || data.link.name || "no title",
        link,
      });
    } else {
      return out;
    }
  };
  return Compiler;
}

export { plugin as dendronNoteRefPlugin };
export { CompilerOpts as DendronNoteRefPluginOpts };
