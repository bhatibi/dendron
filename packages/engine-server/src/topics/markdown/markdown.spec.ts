import { EngineTestUtils } from "@dendronhq/common-server";
import _ from "lodash";
import { replaceRefs } from "./plugins/replaceRefs";
import { getProcessor } from "./utils";
import { DendronEngine } from "../../engine";
import { DNodeUtils, Note } from "@dendronhq/common-all";

// @ts-ignore
const mdSimple = `
# H1 Header

h1 content

h1.1 content

## H2 Header

h2 content

### H3 Header

h3 content

- bullet1
- bullet2

1. ordered1
2. ordered2

\`\`\`
code fence
\`\`\`

\`code span\`

- [link](normal-link)
- ![image](image-link.jpg)
- [[foo-wiki-link]]
- [[label|foo-wiki-link]]
- [[label|foo-wiki-link]]#foobar

HTML

*[HTML]: HyperText Markup Language
`;

describe("replaceRefs", () => {
  // @ts-ignore
  let root: string;

  beforeEach(() => {
    root = EngineTestUtils.setupStoreDir();
  });

  test("imagePrefix", () => {
    // const uri = path.join(root, "sample.image-link.md");
    // const note = createNoteFromMarkdown(uri);
    // expect(note.toRawProps(false, { ignoreNullParent: true })).toMatchSnapshot(
    //   "rawprops"
    // );
    const out = getProcessor()
      .use(replaceRefs, { imageRefPrefix: "bond/", scratch: "" })
      .processSync(`![alt-text](image-url.jpg)`);
    expect(_.trim(out.toString())).toEqual("![alt-text](bond/image-url.jpg)");
  });

  test("wiki2Md", () => {
    const links = `
[link](normal-link)

- [[foo-wiki-link]]
- [[label|foo-wiki-link]]
- [[label|foo-wiki-link]]#foobar
    `;
    const proc = getProcessor().use(replaceRefs, {
      wikiLink2Md: true,
      scratch: "",
    });
    const out = proc.processSync(links);
    const tokens = proc.parse(links);

    expect(out.toString()).toMatchSnapshot("raw");
    expect(tokens).toMatchSnapshot("parsed");
  });

  test("wiki2Md and swap id", async () => {
    const engine = DendronEngine.getOrCreateEngine({ root });
    await engine.init();
    const proc = getProcessor().use(replaceRefs, {
      wikiLink2Md: true,
      wikiLinkUseId: true,
      engine,
      scratch: "",
    });

    const note = DNodeUtils.getNoteByFname(
      "engine-server.replace-refs",
      engine,
      { throwIfEmpty: true }
    ) as Note;
    const out = proc.processSync(note.body);

    expect(out.toString()).toMatchSnapshot("raw");
  });
});

describe("Markdown extensions", () => {
  test("abbreviations are preserved", async () => {
    const abbr = `
HTML, JS and CSS are the fundamental building blocks of the web.

*[CSS]: Cascading Style Sheets
*[HTML]: HyperText Markup Language
*[JS]: JavaScript
    `;
    const proc = getProcessor();
    const out = await proc.process(abbr);
    expect(out.toString()).toMatchSnapshot();
  });
});
