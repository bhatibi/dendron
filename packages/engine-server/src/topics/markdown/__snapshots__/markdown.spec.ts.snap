// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown extensions abbreviations are preserved 1`] = `
"HTML, JS and CSS are the fundamental building blocks of the web.

    

*[HTML]: HyperText Markup Language
*[JS]: JavaScript
*[CSS]: Cascading Style Sheets"
`;

exports[`replaceRefs wiki2Md and swap id: raw 1`] = `
"Replace-refs Text

- [engine-server.replace-refs.one](234)
- [lbl](234)
- [engine-server.replace-refs.one.md](234)
- [lbl](234)

"
`;

exports[`replaceRefs wiki2Md: parsed 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 6,
                  "line": 2,
                  "offset": 6,
                },
                "indent": Array [],
                "start": Object {
                  "column": 2,
                  "line": 2,
                  "offset": 2,
                },
              },
              "type": "text",
              "value": "link",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 20,
              "line": 2,
              "offset": 20,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "title": null,
          "type": "link",
          "url": "normal-link",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 20,
          "line": 2,
          "offset": 20,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "alias": "foo-wiki-link",
                    "exists": false,
                    "hChildren": Array [
                      Object {
                        "type": "text",
                        "value": "foo-wiki-link",
                      },
                    ],
                    "hName": "a",
                    "hProperties": Object {
                      "className": "internal new",
                      "href": "#/page/foo-wiki-link",
                    },
                    "permalink": "foo-wiki-link",
                  },
                  "position": Position {
                    "end": Object {
                      "column": 20,
                      "line": 4,
                      "offset": 41,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 4,
                      "offset": 24,
                    },
                  },
                  "type": "wikiLink",
                  "value": "foo-wiki-link",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 20,
                  "line": 4,
                  "offset": 41,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 4,
                  "offset": 24,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 20,
              "line": 4,
              "offset": 41,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 4,
              "offset": 22,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "alias": "label",
                    "exists": false,
                    "hChildren": Array [
                      Object {
                        "type": "text",
                        "value": "label",
                      },
                    ],
                    "hName": "a",
                    "hProperties": Object {
                      "className": "internal new",
                      "href": "#/page/foo-wiki-link",
                    },
                    "permalink": "foo-wiki-link",
                  },
                  "position": Position {
                    "end": Object {
                      "column": 26,
                      "line": 5,
                      "offset": 67,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 5,
                      "offset": 44,
                    },
                  },
                  "type": "wikiLink",
                  "value": "foo-wiki-link",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 26,
                  "line": 5,
                  "offset": 67,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 5,
                  "offset": 44,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 5,
              "offset": 67,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 5,
              "offset": 42,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "alias": "label",
                    "exists": false,
                    "hChildren": Array [
                      Object {
                        "type": "text",
                        "value": "label",
                      },
                    ],
                    "hName": "a",
                    "hProperties": Object {
                      "className": "internal new",
                      "href": "#/page/foo-wiki-link",
                    },
                    "permalink": "foo-wiki-link",
                  },
                  "position": Position {
                    "end": Object {
                      "column": 26,
                      "line": 6,
                      "offset": 93,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 6,
                      "offset": 70,
                    },
                  },
                  "type": "wikiLink",
                  "value": "foo-wiki-link",
                },
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 5,
                      "line": 7,
                      "offset": 105,
                    },
                    "indent": Array [
                      3,
                    ],
                    "start": Object {
                      "column": 26,
                      "line": 6,
                      "offset": 93,
                    },
                  },
                  "type": "text",
                  "value": "#foobar
  ",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 5,
                  "line": 7,
                  "offset": 105,
                },
                "indent": Array [
                  3,
                ],
                "start": Object {
                  "column": 3,
                  "line": 6,
                  "offset": 70,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 5,
              "line": 7,
              "offset": 105,
            },
            "indent": Array [
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 6,
              "offset": 68,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": Position {
        "end": Object {
          "column": 5,
          "line": 7,
          "offset": 105,
        },
        "indent": Array [
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 4,
          "offset": 22,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 5,
      "line": 7,
      "offset": 105,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`replaceRefs wiki2Md: raw 1`] = `
"[link](normal-link)

- [foo-wiki-link](foo-wiki-link)
- [label](foo-wiki-link)
- [label](foo-wiki-link)#foobar
    

"
`;
